hack/ci-utils/go_generate.sh

```diff
@@ -0,0 +1,64 @@
+#!/bin/bash
+
+set -o errexit
+
+
+main() {
+
+    # Variables checked for timeout values, defaults used if unset
+    soft_timeout="${GO_GENERATE_SOFT_TIMEOUT:-20m}"
+    hard_timeout="${GO_GENERATE_HARD_TIMEOUT:-21m}"
+
+    if [[ -n $CI ]]; then
+        set -x
+    fi
+
+
+    # Declare local before using command substitution so that local doesn't cover up timeout's exit code.
+    local timeout_exitcode go_generate
+    # go generate doesn't support timeout itself, so --preserve-status wouldn't help us as it would simply be the status timeout sent to go.
+    go_generate="go generate ./..."
+    echo "Running ${go_generate} sending SIGTERM after ${soft_timeout} and SIGKILL after ${hard_timeout}"
+    echo "${go_generate}"
+     # shellcheck disable=SC2086
+    timeout --foreground --kill-after="${hard_timeout}" -v -s SIGTERM "${soft_timeout}" ${go_generate}
+    timeout_exitcode=$?
+
+    # Capture expected exit codes
+    case $timeout_exitcode in
+    124)
+        echo "Command ${go_generate} timed out."
+        ;;
+
+    125)
+        echo "The command \"timeout\" failed trying to run ${go_generate}. Note that ${go_generate} did not fail."
+        ;;
+
+    126)
+        echo "Command found but cannot be invoked."
+        ;;
+
+    127)
+        echo "Command not found recieved."
+        ;;
+
+    137)
+        echo "timeout or command recieved kill signal 9) SIGKILL."
+        ;;
+
+    -)
+        echo "Command go generate ./... exited with exit_code code ${-}."
+        ;;
+    0)
+        echo "Command ${go_generate} completed successfully, exit code ${timeout_exitcode}."
+        ;;
+
+    *)
+        echo "Unexpected error code received: ${timeout_exitcode}."
+        ;;
+    esac
+
+    exit "${timeout_exitcode}"
+}
+
+main
```

Comment:
My 0.02 - we can remove this `main` function / cmd since we're not calling this logic anywhere outside of this file.

Reply:
I typically create a main function in most of my bash programs, the reason is it puts the "main" logic at the top of the script. Since bash executes every line sequentially it can be harder to read this way.
I'm happy to pull it out of the function though if you think we should.

Reply:
ah, I'm a little sad to see this removed, honestly... I use the same pattern @s-fairchild, and appreciate seeing someone else using it too! I think `main` in bash looks a lot cleaner and makes the code more organized into `set`s, `main`, and other various helpers.

Reply:
@SudoBrendan There's still time, I can add `main` back :)
It is a good design pattern, even for small scripts like this I think it helps.

Reply:
If anyone still isn't sold, I think the biggest win of `main` is that you don't have to scroll through "how the sausage is made" in your script before you hit the top-level workflow. That is, in bash, a (helper) function has to be defined before it is used. That inevitably leads to you defining a bunch of seemingly "random" helpers until you suddenly have top-level code that uses it, something like:

```
#!/bin/bash

func foo() { ... }

func bar() { ... }

func biz() { ... }

foo()
bar()
biz()
```

With `main`, you can put that all at the top and (hopefully) define easily legible top-level logic flow with some cool function names, and code starts to look like:

```
#!/bin/bash

func main() {
    foo()
    bar()
    biz()
}

func foo() { ... }

func bar() { ... }

func biz() { ... }

main() # kinda icky to have this "floating" top-level call, but hey, it's the best we can do. There's exactly 1 of these calls, and it ideally happens as the last line of every file...
```